src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs

```diff
@@ -1622,29 +1617,43 @@ static bool BindingFlagsAreSupported (BindingFlags bindingFlags)
 			return (bindingFlags & BindingFlags.IgnoreCase) == BindingFlags.IgnoreCase || (int) bindingFlags > 255;
 		}
 
-		void MarkTypeForDynamicallyAccessedMembers (ref ReflectionPatternContext reflectionContext, TypeDefinition typeDefinition, DynamicallyAccessedMemberTypes requiredMemberTypes)
+		void MarkTypeForDynamicallyAccessedMembers (ref ReflectionPatternContext reflectionContext, TypeReference typeReference, DynamicallyAccessedMemberTypes requiredMemberTypes)
 		{
-			foreach (var member in typeDefinition.GetDynamicallyAccessedMembers (requiredMemberTypes)) {
-				switch (member) {
-				case MethodDefinition method:
-					MarkMethod (ref reflectionContext, method);
-					break;
-				case FieldDefinition field:
-					MarkField (ref reflectionContext, field);
-					break;
-				case TypeDefinition nestedType:
-					MarkNestedType (ref reflectionContext, nestedType);
-					break;
-				case PropertyDefinition property:
-					MarkProperty (ref reflectionContext, property);
-					break;
-				case EventDefinition @event:
-					MarkEvent (ref reflectionContext, @event);
-					break;
-				case null:
-					var source = reflectionContext.Source;
-					reflectionContext.RecordRecognizedPattern (typeDefinition, () => _markStep.MarkEntireType (typeDefinition, includeBaseTypes: true, new DependencyInfo (DependencyKind.AccessedViaReflection, source), source));
-					break;
+			var foundType = typeReference?.Resolve ();
+			if (foundType == null) {
+				// Intentionally ignore - it's not wrong for code to call Type.GetType on non-existing name, the code might expect null/exception back.
+				reflectionContext.RecordHandledPattern ();
+				return;
+			} else {
+				foreach (var member in foundType.GetDynamicallyAccessedMembers (requiredMemberTypes)) {
+					switch (member) {
+					case MethodDefinition method:
+						MarkMethod (ref reflectionContext, method);
+						break;
+					case FieldDefinition field:
+						MarkField (ref reflectionContext, field);
+						break;
+					case TypeDefinition nestedType:
+						MarkNestedType (ref reflectionContext, nestedType);
+						break;
+					case PropertyDefinition property:
+						MarkProperty (ref reflectionContext, property);
+						break;
+					case EventDefinition @event:
+						MarkEvent (ref reflectionContext, @event);
+						break;
+					case null:
+						var source = reflectionContext.Source;
+						reflectionContext.RecordRecognizedPattern (foundType, () => _markStep.MarkEntireType (foundType, includeBaseTypes: true, new DependencyInfo (DependencyKind.AccessedViaReflection, source), source));
+						break;
+					}
+				}
+			}
+
+			// If this is a generic type, mark all its arguments.
+			if (typeReference is GenericInstanceType genericType) {
+				foreach (var arg in genericType.GenericArguments) {
+					MarkTypeForDynamicallyAccessedMembers (ref reflectionContext, arg, requiredMemberTypes);
```

Comment:
This is excessive and sort of wrong.

Take for example:
```C#
class MyType
{
    public void UnusedMethod() {}
}

void Test()
{
    Type listType = Type.GetType("System.Collections.Generic.List`1[MyType]");
    MethodInfo addMethod = listType.GetMethod("Add");
}
```

The call to `GetMethod` is annotated with `DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)`. So with your code it would mark `List`1.Add` but also `MyType.UnusedMethod`.

I think the easiest way to fix this would be to simply call `MarkType` on the `TypeReference` and then apply additional marking based on the annotations. `MarkType` on its own should not include more then necessary for the type (on simply cases it will not mark any methods and so on).

For non-generic cases this is already happening, since for example marking a method on the type (due to annotations) will also mark the type (MarkMethod does that). The problem is with generics since by calling `.Resolve` we lose the generic instantiation part.

Reply:
In a way this is a very similar problem to https://github.com/mono/linker/issues/1559.
Can you please validate if your change also affects the behavior of arrays as noted in #1559?
Either make sure to note the differences in the other issue, or maybe try to fix both at once - I would expect the fixes to be relatively similar.
Basically we need to pass around `TypeReference` instead of `TypeDefinition`.

Reply:
Makes sense. I wasn't sure how much to keep of a type that is passed as generic argument in a dynamically accessed member - I changed things so that we use `MarkTypeVisibleToReflection` to mark the type references of these.

Based on the discussion in #1559 I changed `ProcessGenericArgumentDataFlow` so that we mark the element type of the generic argument that is passed. This way, passing `T[]` will result in the linker marking `T`.

Reply:
The discussion in #1559 is at best inconclusive. I think we definitely need to mark the array type itself (based on the annotations). Right now we're also marking the element type based on the annotations, which is technically incorrect. I'm starting to agree with @MichalStrehovsky that we can stop doing that - the cases where this would matter should be very few.

Note that the original repro with JSON serializing `T[]` is sort of "wrong" anyway. We don't have an annotation which would work for serializers - it would have to be something like `RecursivePublicProperties`. Once we have that, it could also handle arrays (as in mark the element types recursively). Something like this has been proposed a while ago: https://github.com/mono/linker/issues/1087.